index.js.flow

```diff
@@ -3,22 +3,30 @@
 opaque type Next = Function | void;
 opaque type Yield = mixed;
 
-export type Gensync<T> = {
-  (...args: Array<mixed>): Handler<T>,
-  sync(...args: Array<mixed>): T,
-  async(...args: Array<mixed>): Promise<T>,
-  errback(...args: Array<mixed>): void,
+export type Gensync<Args, Return> = {
+  (...args: Args): Handler<Return>,
+  sync(...args: Args): Return,
+  async(...args: Args): Promise<Return>,
+  // ...args: [...Args, Callback]
+  errback(...args: any[]): void,
 };
 
-export type Handler<T> = Generator<Yield, T, Next>;
-export type Options<T> = {
-  sync: (...args: Array<mixed>) => T,
-  async: (...args: Array<mixed>) => Promise<T>,
-  errback: (...args: Array<mixed>) => void,
+export type Handler<Return> = Generator<Yield, Return, Next>;
+export type Options<Args, Return> = {
+  sync(...args: Args): Return,
   arity?: number,
   name?: string,
-};
+} & (
+  | { async(...args: Args): Promise<Return> }
+  // ...args: [...Args, Callback]
+  | { errback(...args: any[]): void }
```

Comment:
Technically these are both optional too. Does `| {}` also work here? I have no idea if Flow can handle that.

Reply:
This allows either one or the other, but warns of both are used.
By optional, do you mean that it's possible to specify neither?

Reply:
At the moment it uses the sync function if neither are provided.

Reply:
`| {}` doesn't work because it isn't an exact object, and if it was we couldn't use `&`.